### PR TITLE
Try: WooCommerce User Meta

### PR DESCRIPTION
**Pull request by emdashcodes**
This PR adds the ability to get user information, specifically meta, and update it from wc-admin.

I started looking into this as a part of #178, but it will be needed for persisting other user settings like https://github.com/woocommerce/wc-admin/issues/114.

There are more improvements that could be made here. For example, we could preload the user data on initial load with PHP. It looks like Gutenberg has some helpers we could use for this. There is also no error handling if the requests fail. I wasn't sure what we want to do in this case (or other critical requests) so I figured we could have a discussion about that and follow-up. I wanted to get a first-pass up in the meantime.

This PR specifically introduces an inbox last read time meta key. The integration there is basic since we are still deciding the future of the feature, but it works for saving and retrieving.

To Test:
* `npm install`
* `npm test` to make sure all tests pass
* `npm start`
* Open the inbox, verify the top notice is marked as unread.
* Switch to a different tab, then back to inbox. The orange dot should be gone.
* You can also monitor the Redux dev console, looking for `SET_USER`. You should see two requests. Make sure to select `wc-admin` in the instance selector (it defaults to core data). You can also verify the POST request worked via the Network console.

### CODE REVIEW

File: client/layout/activity-panel/panels/inbox.js [hunk 1]
```diff
@@ -70,14 +72,26 @@ class InboxPanel extends Component {
...
 			notices: [],
+			lastRead: null,
 		};
 	}
 
 	componentDidMount() {
+		// @TODO Temporary
+		// When we start pulling reviews from the actual API, we'll want to show a loading indicator
+		// while both results load (using isRequestingCurrentUser and a isRequestinReviews), and also wait to
+		// mark the last opened date until after review results load.
```
Comment (ryelle) [inline client/layout/activity-panel/panels/inbox.js]: Nitpick, but this is the inbox panel, not reviews.

File: client/store/current-user/actions.js [hunk 2]
```diff
@@ -0,0 +1,33 @@
+/** @format */
+
+/**
+ * External dependencies
+ */
+import { dispatch } from '@wordpress/data';
+import apiRequest from '@wordpress/api-request';
```
Comment (ryelle) [inline client/store/current-user/actions.js]: [`api-request` has been deprecated](https://github.com/WordPress/gutenberg/commit/247a6e07c22792769c4d397ad3a6d118dbc90a77) in favor of [api-fetch](https://github.com/WordPress/gutenberg/tree/master/packages/api-fetch) (though you can also POST with fetch)

File: client/store/current-user/actions.js [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+			user,
+		};
+	},
+	updateWoocommerceUserMeta( meta ) {
+		return async () => {
+			const data = {
+				woocommerce_meta: { ...meta },
+			};
+			try {
+				const updatedUser = await apiRequest( {
```
Comment (ryelle) [inline client/store/current-user/actions.js]: It seems like core doesn't use this approach of putting api requests inside actions, rather [their actions are all plain objects](https://github.com/WordPress/gutenberg/blob/master/packages/core-data/src/actions.js), and it looks like instead they have [a middleware for effects](https://github.com/WordPress/gutenberg/blob/master/packages/editor/src/store/effects.js), where post/delete/put requests happen. [This `trashPost` function](https://github.com/WordPress/gutenberg/blob/master/packages/editor/src/store/effects/posts.js#L268), which is triggered [on the TRASH_POST action](https://github.com/WordPress/gutenberg/blob/master/packages/editor/src/store/effects.js#L87) seems like a relatively simple example, though I'm not sure how we can hook into that (maybe we also set up [a middlewares file](https://github.com/WordPress/gutenberg/blob/master/packages/editor/src/store/middlewares.js)). `edit-post` also has this `effects` pattern, and I don't think anything else in gutenberg updates data.

**Review by ryelle — COMMENTED** [review]
Everything works as you described (though the unread indicator on the Inbox tab itself doesn't go away). I left some comments about core patterns - we should all try to get on the same page as gutenberg before we get too far into data work.

File: client/store/current-user/resolvers.js [hunk 4]
```diff
@@ -0,0 +1,18 @@
+/** @format */
+/**
+ * External dependencies
+ */
+import { dispatch } from '@wordpress/data';
+import apiRequest from '@wordpress/api-request';
+
+export default {
+	async getCurrentUser() {
+		const path = '/wp/v2/users/me?context=edit';
```
Comment (timmyc) [inline client/store/current-user/resolvers.js]: For some reason this is 404'ing for me:

image

**Review by timmyc — COMMENTED** [review]
Fantastic Justin! Thanks for diving in the deep end on this one. I really like what @ryelle mentioned in her review around trying to match the patterns in gutenberg here. Loving all the test coverage too.

I noted that I was seeing a 404 on one request, otherwise things are working well for me.

File: client/layout/activity-panel/panels/inbox.js [hunk 5]
```diff
@@ -4,8 +4,9 @@
  */
 import { __ } from '@wordpress/i18n';
 import { Button } from '@wordpress/components';
-import { Component, Fragment } from '@wordpress/element';
+import { Component, Fragment, compose } from '@wordpress/element';
```
Comment (psealock) [inline client/layout/activity-panel/panels/inbox.js]: I'm getting a warning in the console, although this problem is surely in several places

>wp.element.compose is deprecated and will be removed in 3.5. Please use wp.compose.compose instead.

File: client/store/current-user/test/selectors.js [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+		state = deepFreeze( {
+			currentUser: {
+				data: {
+					first_name: 'Testing',
+				},
+			},
+		} );
+		expect( getCurrentUser( state ) ).toEqual( {
+			first_name: 'Testing',
+		} );
```
Comment (psealock) [inline client/store/current-user/test/selectors.js]: I'm getting an error with this one, although the tests pass

screen shot 2018-08-06 at 12 02 10 pm

**Comment by justinshreve** [comment]
Thanks for the 👀everyone. Marking this In Progress for now since it's not part of the current sprint, and I'm AFK a couple days this week, but I hope I can circle back next sprint/next week to get this updated - specifically around things like the Gutenberg effect patterns.

**Comment by ryelle** [comment]
@justinshreve When you're back from AFK you might want to check out [@wordpress/redux-routine](https://github.com/WordPress/gutenberg/tree/master/packages/redux-routine) - it should let you create a redux middleware to send off the API requests in response to actions. It's very new, not on npm yet, but probably a good place to watch for how plugins should use data.

**Comment by justinshreve** [comment]
I'm closing this PR out for now until we get back to the user meta task. We can keep the branch around for now.

**Pull request closed without merging**